### PR TITLE
Feature/two factor auth

### PR DESCRIPTION
**Pull request by VAFomin**
- Added two-factor authentication with app

### CODE REVIEW

File: packages/asc-web-common/api/settings/index.js [hunk 1]
```diff
@@ -249,3 +249,69 @@ export function updateConsumerProps(newProps) {
...
+    method: "get",
+    url: "/settings/tfaapp/setup",
+  };
+
+  if (confirmKey) options.headers = { confirm: confirmKey };
+
+  return request(options);
+}
+
+export function loginWithTfaCodeAndCookie(code) {
```
Comment (AlexeySafronov) [inline packages/asc-web-common/api/settings/index.js:307]: Rename to validateTfaCode

File: packages/asc-web-common/store/AuthStore.js [hunk 2]
```diff
@@ -153,18 +156,33 @@ class AuthStore {
     try {
       const response = await api.user.login(user, hash);
 
-      if (!response || !response.token) throw "Empty API response";
+      if (!response || (!response.token && !response.tfa))
+        throw "Empty API response";
```
Comment (AlexeySafronov) [inline packages/asc-web-common/store/AuthStore.js:160]: Added translation

File: packages/asc-web-common/store/TfaStore.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    return type;
+  };
+
+  setTfaSettings = async (type) => {
+    return await api.settings.setTfaSettings(type);
+  };
+
+  getTfaConfirmLink = async (res, type) => {
+    if (res && type !== "none") {
```
Comment (AlexeySafronov) [inline packages/asc-web-common/store/TfaStore.js:33]: Simplify condition

File: products/ASC.People/Client/public/locales/en/Profile.json [hunk 4]
```diff
@@ -59,5 +57,10 @@
   "Disconnect": "Disconnect",
 
   "ProviderSuccessfullyConnected": "Provider successfully connected",
-  "ProviderSuccessfullyDisconnected": "Provider successfully disconnected"
+  "ProviderSuccessfullyDisconnected": "Provider successfully disconnected",
+  "TwoFactorDescription": "<strong>Two-factor authentication</strong> via code generating application was enabled for all users by cloud service administrator.",
```
Comment (AlexeySafronov) [inline products/ASC.People/Client/public/locales/en/Profile.json:61]: Remove html tags from translation

File: products/ASC.People/Client/public/locales/ru/BackupCodesDialog.json [hunk 5]
```diff
@@ -0,0 +1,9 @@
+{
+  "BackupCodesTitle": "Резервные коды",
+  "BackupCodesDescription": "Резервные коды позволяют подтвердить вход, когда у вас нет доступа к телефону.",
+  "BackupCodesSecondDescription": "Каждый код можно использовать только один раз. Вы можете получить новые коды в любое время. Действительны только те коды, которые были сгенерированы последними.",
+  "CodesCounter": "<strong>{{count}} кодов:</strong>",
```
Comment (AlexeySafronov) [inline products/ASC.People/Client/public/locales/ru/BackupCodesDialog.json:5]: Remove html tags from translation

File: products/ASC.People/Client/public/locales/ru/Profile.json [hunk 6]
```diff
@@ -59,5 +58,10 @@
   "Disconnect": "Отключить",
 
   "ProviderSuccessfullyConnected": "Провайдер успешно подключен",
-  "ProviderSuccessfullyDisconnected": "Провайдер успешно отключен"
+  "ProviderSuccessfullyDisconnected": "Провайдер успешно отключен",
+  "TwoFactorDescription": "<strong>Двухфакторная аутентификация</strong> через приложение для генерации кодов включена для всех пользователей администратором облачного сервиса.",
```
Comment (AlexeySafronov) [inline products/ASC.People/Client/public/locales/ru/Profile.json:62]: Remove html tags from translation

File: products/ASC.People/Client/src/pages/Profile/Section/Body/index.js [hunk 7]
```diff
@@ -123,9 +150,26 @@ class SectionBodyContent extends React.PureComponent {
       console.error(e);
     }
 
+    await getTfaSettings().then((type) => this.setState({ tfa: type }));
```
Comment (AlexeySafronov) [inline products/ASC.People/Client/src/pages/Profile/Section/Body/index.js:153]: Replace "then" to async...await style

File: products/ASC.People/Client/src/pages/Profile/Section/Body/index.js [hunk 8]
```diff
@@ -123,9 +150,26 @@ class SectionBodyContent extends React.PureComponent {
...
+    await getBackupCodes().then((codes) =>
```
Comment (AlexeySafronov) [inline products/ASC.People/Client/src/pages/Profile/Section/Body/index.js:154]: Replace "then" to async...await style

File: products/ASC.People/Client/src/pages/Profile/Section/Body/index.js [hunk 9]
```diff
@@ -123,9 +150,26 @@ class SectionBodyContent extends React.PureComponent {
...
     window.loginCallback = this.loginCallback;
   }
 
+  async componentDidUpdate(prevState) {
+    const { getBackupCodes } = this.props;
+
+    if (
+      this.state.backupCodesDialogVisible !== prevState.backupCodesDialogVisible
+    ) {
+      await getBackupCodes().then((codes) =>
```
Comment (AlexeySafronov) [inline products/ASC.People/Client/src/pages/Profile/Section/Body/index.js:167]: Replace "then" to async...await style

File: web/ASC.Web.Client/public/locales/en/Confirm.json [hunk 10]
```diff
@@ -26,12 +26,18 @@
...
   "ProviderLoginError": "Authorization error",
-  "ProviderNotConnected": "Provider is not connected to your account"
+  "ProviderNotConnected": "Provider is not connected to your account",
+  "EnterAppCodeTitle": "Enter code from authentication app",
+  "EnterAppCodeDescription": "Enter 6-digit generated code from your application. If you don't have access to your phone you can use your backup codes.",
+  "EnterCodePlaceholder": "Enter code",
+  "Continue": "Continue",
+  "SetAppTitle": "Configure your authenticator application",
+  "SetAppDescription": "The two-factor authentication is enabled to provide additional portal security. Configure your authenticator application to continue work on the portal. For example you could use Google Authenticator for <1>Android</1> and <4>iOS</4> or Authenticator for <8>Windows Phone</8>.",
+  "SetAppInstallDescription": "To connect your application scan the QR code or manually enter your secret key <strong>{{key}}</strong> then enter 6-digit code from your application in the field below.",
```
Comment (AlexeySafronov) [inline web/ASC.Web.Client/public/locales/en/Confirm.json:41]: Remove html tags from translation

File: web/ASC.Web.Client/public/locales/ru/Confirm.json [hunk 11]
```diff
@@ -21,12 +21,18 @@
...
   "ProviderLoginError": "Ошибка авторизации",
-  "ProviderNotConnected": "Провайдер не подключен к вашему аккаунту"
+  "ProviderNotConnected": "Провайдер не подключен к вашему аккаунту",
+  "EnterAppCodeTitle": "Введите код из приложения аутентификации",
+  "EnterAppCodeDescription": "Введите 6-значный сгенерированный код из вашего приложения. Если у вас нет доступа к телефону, вы можете использовать резервные коды.",
+  "EnterCodePlaceholder": "Введите код",
+  "Continue": "Продолжить",
+  "SetAppTitle": "Настройте приложение для аутентификации",
+  "SetAppDescription": "Для повышения безопасности на портале включена двухфакторная аутентификация. Чтобы продолжить работу на портале, настройте приложение для аутентификации. Например, вы можете использовать Google Authenticator для <1>Android</1> и <4>iOS</4> или Authenticator для <8>Windows Phone</8>.",
+  "SetAppInstallDescription": "Для подключения приложения отсканируйте QR-код или вручную введите секретный ключ <strong>{{key}}</strong>, затем введите 6-значный код из приложения в поле ниже.",
```
Comment (AlexeySafronov) [inline web/ASC.Web.Client/public/locales/ru/Confirm.json:36]: Remove html tags from translation

File: web/ASC.Web.Client/src/components/pages/Confirm/index.js [hunk 12]
```diff
@@ -57,6 +59,18 @@ const Confirm = ({ match }) => {
         path={`${match.path}/PortalOwnerChange`}
```
Comment (AlexeySafronov) [inline web/ASC.Web.Client/src/components/pages/Confirm/index.js:59]: Extract path to separated const

File: web/ASC.Web.Client/src/components/pages/Settings/categories/security/accessPortal.js [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+          orientation="vertical"
+          options={[
+            {
+              label: t("Disabled"),
+              value: "none",
+            },
+            {
+              label: t("BySms"),
+              value: "sms",
+              disabled: true, //TODO: get from api, when added sms tfa
```
Comment (AlexeySafronov) [inline web/ASC.Web.Client/src/components/pages/Settings/categories/security/accessPortal.js:103]: Fix todo

**Review by ilyaoleshko — APPROVED** [review]

**Review by AlexeySafronov — APPROVED** [review]

**Pull request merged**